docs/dev/data_validation.rst

```diff
@@ -0,0 +1,688 @@
+   `Dagster's asset checks <https://docs.dagster.io/guides/test/asset-checks>`_.
+   Typically these tests weren't well suited to SQL, weren't performance bottlenecks,
+   and had already been implemented in Python. E.g. :func:`pudl.validate.no_null_rows`.
+
+The ``dbt_helper`` script
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We've created a script that helps make working with the dbt tests within PUDL a little
+more ergonomic. It's called :mod:`pudl.scripts.dbt_helper` and is installed in the
+``pudl-dev`` conda environment. ``dbt_helper validate`` runs the data validation tests
+and provides richer output when a test fails than ``dbt build``. It also allows us to
+use the `Dagster asset selection syntax
+<https://docs.dagster.io/guides/build/assets/asset-selection-syntax/reference>`_.
+
+Example usage:
+
+.. code-block:: bash
+
+    # for just a single asset
+    dbt_helper validate --asset-select "key:out_eia__yearly_generators"
+    # for this asset as well as all upstream assets
+    dbt_helper validate --asset-select "+key:out_eia__yearly_generators"
+    # same as above, but skip row counts
+    dbt_helper validate --asset-select "+key:out_eia__yearly_generators" --exclude "*check_row_counts*"
+
+See ``dbt_helper validate --help`` for usage details.
+
+Data validation in our integration tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The dbt data tests are invoked by ``pytest`` as part of our integration tests. This
+means they run as part of our continuous integration (CI) checks before a PR can be
+merged into ``main``. However, the CI only processes 1-2 years of data, so when the
+tests run in CI, they're only checking a small subset of the data we publish. We also
+don't run the row count checks in CI, since the fast ETL outputs are more changeable
+and less informative than those in the nightly builds.
+
+This means that when you're developing a new table or updating an existing table, it's
+important to manually run the dbt tests on the new data in its entirety before the
+changes are merged into ``main``.
+
+If the data validations fail in the ``pytest`` integration tests, they should produce
+helpful output indicating what failed and why, in the same way as ``dbt_helper
+validate``
+
+Data validation in branch builds
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Depending on your computer, running the full ETL locally can be extremely time consuming
+and may run into memory limits. It's also easy to accidentally end up with local outputs
+that are the result of code from multiple different branches, and so may not be
+consistent with each other. If you're only altering a few tables, rematerializing them
+in Dagster and then running the specific dbt tests that apply to them and any tables
+downstream of them should work fine.
+
+Kicking off a branch build
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+But especially when we're doing big quarterly or annual updates, and dozens or hundreds
```

Comment:
I see how this was meant to be a transition from the previous section, but I don't think we really need it -- and without it, the sentence will make more sense to someone jumping to this section directly.

```suggestion
When we're doing big quarterly or annual updates, and dozens or hundreds
```